proposed/MPID Messaging.md

```diff
@@ -0,0 +1,451 @@
+
+We also have identified a need for some form of secure messaging in order to implement safecoin, so failure to implement this RFC would impact on that too.
+
+# Unresolved Questions
+
+1. `MpidMessage` and `MpidHeader` are wrapped in `StructuredData` instances in order to allow Delete requests on them and to allow them to be given different tags (otherwise they could be ImmutableData, since they don't mutate).  Should Routing, which already has to define these two types, be made "aware" of these two types?  (i.e. should they get added to the [`::routing::data::Data` enum][6])
+
+    Identified pros:
+    - less wrapping/parsing, so simpler for Vaults and Clients to deal with
+    - more efficient (smaller messages to transmit and store)
+    - no need for Routing to publicly expose `MPID_MESSAGE_TAG` or `MPID_HEADER_TAG`
+
+    Identified cons:
+    - increased complexity in Routing
+    - would need to add the sender's public key to the header Put flow
+
+    Qi prefers using `StructuredData`, Fraser prefers not using `StructuredData` unless the effort required by Routing to accommodate the new types is significant.
+
+# Future Work
+
+- This RFC doesn't address how Clients get to "know about" each other.  Future work will include details of how Clients can exchange MPIDs in order to build an address book of peers.
+
+- It might be required to provide Vault-to-Vault, Client-to-Vault or Vault-to-Client communications in the future.  Potential use cases for this are:
+
+    1. Vault-to-Client notification of a successful safecoin mining attempt
+    1. Client-to-Vault request to take ownership of the Vault
+    1. Vault-to-Client notification of low resources on the Vault
+
+    In this case, the existing library infrastructure would probably need significant changes to allow a Vault to act as an MPID Client (e.g. the MPID struct is defined in the SAFE Client library).
+
+- Another point is that (as with MAID accounts), there is no cleanup done by the network of MpidMessages if the user decides to stop using SAFE.
+
+- Also, not exactly within the scope of this RFC, but related to it; MPID packets at the moment have no human-readable name.  It would be more user-friendly to provide this functionality.
+
+# Appendix
+
+### Further Implementation Details
+
+All MPID-related messages will be in the form of a Put, Post or Delete of a `StructuredData`.
+
+This will be:
+
+```rust
+
+StructuredData {
+    type_tag: MPID_MESSAGE,
+    identifier: mpid_message_name(mpid_message),  // or mpid_header_name(signed_header)
+    data: XXX,
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![sender_public_key],
+    previous_owner_signatures: vec![],
+}
+
+```
+
+where XXX is indicated in the following list of all MPID-related messages:
```

Comment:
If the data field can be a serialised MpidMessageWrapper, MpidMessage, or MpidHeader, then they can not be combined into one request type, otherwise the parser will have issue unless we capture failed parse attempt and try different type one by one.
i.e. if say in a post request, the sd.data maybe a MpidMessage or a MpidHeader, then we have to do something like `if parse<MpidHeader>(sd.data) {} else { parse<MpidMessage>(sd.data); }`.